### ISSUE TITLE
Missing type inference on function arguments

### ISSUE THREAD

**Issue by nicolasstucki** [lampepfl/dotty#2418]
https://github.com/lampepfl/dotty/blob/master/tests/neg/i1802.scala should be a `pos` test, there is a type that gets infered to `Nothing` where the type should be inferred and is inferred in scalac.

The minimized version this issue is:
```scala
object Exception {
  def mkCatcher[Ex](isDef: Ex => Boolean) = ???
  def mkThrowableCatcher(isDef: Throwable => Boolean) = mkCatcher(isDef)
}
```

With `dotc -Xprint:frontend`:
```scala
package <empty> {
  import scala.reflect.ClassTag
  final lazy module val Exception: Exception$ = new Exception$()
  final module class Exception$() extends Object() { this: Exception.type => 
    def mkCatcher[Ex](isDef: Function1[Ex, Boolean]): Nothing = ???
    def mkThrowableCatcher(isDef: Function1[Throwable, Boolean]): Nothing = 
      Exception.mkCatcher[Nothing^](isDef)
  }
}
```
Note the `Nothing` in `Exception.mkCatcher[Nothing^](isDef)`. This is what causes the `Nothing` classtag in `neg/ i1802.scala` and `scala.util.control.Exception`.

With `scalac -Xprint:typer`
```scala
package <empty> {
  object Exception extends scala.AnyRef {
    def <init>(): Exception.type = {
      Exception.super.<init>();
      ()
    };
    def mkCatcher[Ex](isDef: Ex => Boolean): Nothing = scala.this.Predef.???;
    def mkThrowableCatcher(isDef: Throwable => Boolean): Nothing = Exception.this.mkCatcher[Throwable](isDef)
  }
}
```

**Comment by nicolasstucki** [comment]
We should also be able to whitelist `scala.util.control.Exception` if this is fixed.

**Comment by nicolasstucki** [comment]
This would probably not be an issue in this case if #2419 worked like in scalac.

**Comment by odersky** [comment]
scalac and dotty disagree when it comes to the precise rules for inferring Nothing. I don't see a bug here. The error messages for neg/1802 are correct.

**Comment by nicolasstucki** [comment]
Fixed by #3834
